File: src/main/java/dev/ambryn/task/models/Task.java
package dev.ambryn.task.models;

import com.fasterxml.jackson.databind.annotation.JsonSerialize;
import dev.ambryn.task.interfaces.State;
import lombok.Data;

import java.time.LocalDateTime;
import java.time.temporal.ChronoUnit;

@Data
public class Task {
    private static Integer numberOfTasks = 0;
    private Integer id;
    private String name;
    private LocalDateTime createdAt;
    private LocalDateTime startedAt;
    private LocalDateTime finishedAt;
    private int nth;
    private Long result;
    @JsonSerialize(using = StateSerializer.class)
    private State state;

    public Task() {
        this.id = ++numberOfTasks;
        this.createdAt = LocalDateTime.now();
        this.state = new NotStarted(this);
        this.state.setContext(this);
    }

    public Task(String name) {
        this();
        this.name = name;
        this.nth = 40;
    }

    public void start() {
        this.state.start();
    }

    public void finish() {
        this.state.finish();
    }

    public void cancel() {
        this.state.cancel();
    }

    public void error() {
        this.state.error();
    }

    public void describe() {
        System.out.println("Task " + this.id + " - Took " + ChronoUnit.SECONDS.between(startedAt,
                                                                                       finishedAt) + "s to finish! - "
                                   + "Result: " + result);
    }

    public int getNth() {
        return this.nth;
    }

    public void setResult(Long result) {
        this.result = result;
    }
}

File: src/main/java/dev/ambryn/task/models/Task.kt
package dev.ambryn.task.models

import com.fasterxml.jackson.databind.annotation.JsonSerialize
import dev.ambryn.task.interfaces.State
import java.time.LocalDateTime
import java.time.temporal.ChronoUnit

class Task() {
    var name: String = ""
    val id: Int = ++numberOfTasks
    val createdAt: LocalDateTime = LocalDateTime.now()
    var startedAt: LocalDateTime? = null
    var finishedAt: LocalDateTime? = null
    var nth = 40
    var result: Long? = null

    constructor(name: String, nth: Int) : this() {
        this.name = name
        this.nth = nth
    }

    @JsonSerialize(using = StateSerializer::class)
    var state: State = NotStarted(this)

    fun start() {
        state.start()
    }

    fun finish() {
        state.finish()
    }

    fun cancel() {
        state.cancel()
    }

    fun error() {
        state.error()
    }

    fun describe() {
        println(
            "Task $id - Took " + ChronoUnit.SECONDS.between(
                startedAt,
                finishedAt
            ) + "s to finish! - "
                    + "Result: " + result
        )
    }

    companion object {
        private var numberOfTasks = 0
    }
}
